dataplane/src/pipeline.rs

```diff
@@ -1,173 +1,168 @@
+    fn process<Input: Iterator<Item = Packet<Buf>>>(
+        &mut self,
+        input: Input,
+    ) -> impl Iterator<Item = Packet<Buf>> {
+        input.map(|mut packet| {
+            match packet.try_eth_mut() {
+                None => {}
+                Some(mac) => {
+                    mac.set_destination(DestinationMac::new(Mac::BROADCAST).unwrap());
+                }
+            }
+            packet
+        })
     }
+}
 
-    pub fn update_config(&mut self, config: &Config) -> Result<(), Box<dyn Error>> {
-        let result: Result<Vec<_>, _> = self
-            .stages
-            .iter_mut()
-            .map(|stage| stage.update_config(config))
-            .collect();
-        result.map(|_| ())
-    }
+pub struct DecrementTtl;
 
-    pub fn process_packets(
+impl<Buf: PacketBufferMut> NetworkFunction<Buf> for DecrementTtl {
+    fn process<Input: Iterator<Item = Packet<Buf>>>(
         &mut self,
-        packets: Box<dyn Iterator<Item = MetaPacket>>,
-    ) -> Box<dyn Iterator<Item = MetaPacket>> {
-        self.stages
-            .iter_mut()
-            .fold(packets, |packets, stage| stage.process(packets))
+        input: Input,
+    ) -> impl Iterator<Item = Packet<Buf>> {
+        input.filter_map(|mut packet| {
+            match packet.try_ipv4_mut() {
+                None => {}
+                Some(ipv4) => match ipv4.decrement_ttl() {
+                    Ok(()) => return Some(packet),
+                    Err(e) => {
+                        trace!("{e:?}");
+                    }
+                },
+            }
+
+            match packet.try_ipv6_mut() {
+                None => {}
+                Some(ipv6) => match ipv6.decrement_hop_limit() {
+                    Ok(()) => return Some(packet),
+                    Err(e) => {
+                        trace!("{e:?}");
+                    }
+                },
+            }
+
+            None
+        })
     }
 }
 
-#[non_exhaustive]
-pub struct Passthrough;
-
-impl Passthrough {
-    pub fn new() -> Self {
-        Self {}
-    }
+pub struct Chain<Buf: PacketBufferMut, NF1: NetworkFunction<Buf>, NF2: NetworkFunction<Buf>> {
+    nf1: NF1,
+    nf2: NF2,
+    _marker: PhantomData<Buf>,
 }
 
-impl PipelineStage for Passthrough {
-    fn start(&mut self) -> Result<(), Box<dyn Error>> {
-        // nothing to do
-        Ok(())
+impl<Buf: PacketBufferMut, NF1: NetworkFunction<Buf>, NF2: NetworkFunction<Buf>>
+    NetworkFunction<Buf> for Chain<Buf, NF1, NF2>
+{
+    fn process<Input: Iterator<Item = Packet<Buf>>>(
+        &mut self,
+        input: Input,
+    ) -> impl Iterator<Item = Packet<Buf>> {
+        self.nf2.process(self.nf1.process(input))
     }
+}
 
-    fn update_config(&mut self, _: &Config) -> Result<(), Box<dyn Error>> {
-        Ok(())
-    }
+pub trait Pipeline<Buf: PacketBufferMut>: NetworkFunction<Buf> {
+    fn add_stage<NF: NetworkFunction<Buf>>(self, nf: NF) -> impl NetworkFunction<Buf>;
+}
```

Comment:
This is not “Introduce PacketBuffer [...] to dataplane”, this is a complete rewrite of the pipeline architecture.

Reply:
True.  That said, this should be deleted now